### PR TITLE
Add end-to-end tests, make cgo work again, remove cruft from Go distribution.

### PR DESCRIPTION
**Pull request by broady**
Updates #51, #45.

### CODE REVIEW

**Comment by broady** [comment]
Removes about 150 MB from the Go distro (mostly race object files) but adds back gcc and C header files to get cgo to work again.

```
$ docker history gcr.io/google_appengine/golang
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
8b58282ecf0e        50 seconds ago      /bin/sh -c #(nop) CMD ["go-wrapper" "run"]      0 B
aba18eb1758a        50 seconds ago      /bin/sh -c #(nop) WORKDIR /go/src/app           0 B
1ece3012db76        50 seconds ago      /bin/sh -c ln -s /go/src/app /app               11 B
7f5d1fe0c472        51 seconds ago      /bin/sh -c curl https://raw.githubusercontent   2.481 kB
89d292a38d90        51 seconds ago      /bin/sh -c mkdir -p /go/src/app /go/bin && ch   0 B
c7d447d4e02c        52 seconds ago      /bin/sh -c #(nop) ENV GOPATH=/go:/go/src/app/   0 B
cad8ff204d47        52 seconds ago      /bin/sh -c #(nop) ENV PATH=/go/bin:/usr/local   0 B
feba4cd436bb        53 seconds ago      /bin/sh -c curl -sSL https://storage.googleap   113.3 MB
3a3d22e2ab45        12 minutes ago      /bin/sh -c #(nop) ENV GO_WRAPPER_COMMIT=3cdd8   0 B
18a1376abc04        12 minutes ago      /bin/sh -c #(nop) ENV GO_VERSION=1.6.3          0 B
33f8989c0c3a        12 minutes ago      /bin/sh -c apt-get update     && apt-get inst   94.67 MB
5d735d16c1df        3 weeks ago         /bin/sh -c apt-get -q update &&     apt-get i   7.911 MB
193fa0359d01        3 weeks ago         /bin/sh -c #(nop) ENV PORT=8080                 0 B
e8eb529bd074        3 weeks ago         /bin/sh -c #(nop) ENV DEBIAN_FRONTEND=noninte   0 B
abf0692dab5b        3 weeks ago         /bin/sh -c #(nop) CMD []                        0 B
b21b3e3bc691        8 weeks ago         /bin/sh -c #(nop) ADD file:8086d9d18e7f7c09e5   123.6 MB
b29ae6f49e5a        8 weeks ago         /bin/sh -c #(nop) ARG DEBIAN_SUITE              0 B
3bc7cdb36864        12 months ago       /bin/sh -c #(nop) ENV PATH=/usr/local/sbin:/u   0 B
4bfc20f0ed66        12 months ago       /bin/sh -c #(nop) ENV DEBIAN_FRONTEND=noninte   0 B
```

File: base/Dockerfile [hunk 1]
```diff
@@ -28,7 +28,27 @@ RUN curl -sSL https://storage.googleapis.com/golang/go$GO_VERSION.linux-amd64.ta
     curl -sSL https://storage.googleapis.com/golang/go$GO_VERSION.linux-amd64.tar.gz.sha256 -o /tmp/go.tar.gz.sha256 && \
     echo "$(cat /tmp/go.tar.gz.sha256)  /tmp/go.tar.gz" | sha256sum -c - && \
     tar -C /usr/local -vxzf /tmp/go.tar.gz && \
-    rm /tmp/go.tar.gz /tmp/go.tar.gz.sha256
+    rm /tmp/go.tar.gz /tmp/go.tar.gz.sha256 && \
+    # Remove unnecssary files from the Go distro.
```
Comment (zombiezen) [inline base/Dockerfile]: typo: unnecessary

Any rhyme or reason to how we can trim these in general?  Was this just manual inspection?

File: testdata/simple/main.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2016 Google Inc. All rights reserved.
```
Comment (nkubala) [inline testdata/simple/main.go]: what conditions are you looking for to check is cgo is set up? e.g. binaries existing in some directory, making sure a binary is on the path, running a command to check the version, etc.

Comment (zombiezen) [inline testdata/simple/main.go]: Slightly more complex: the proper way to test that we have things set up is to just compile and run a small program that makes calls to C (think autoconf's helper programs).  We could check for individual binaries, but it's not necessarily a good indicator that cgo will work.

Comment (nkubala) [inline testdata/simple/main.go]: ah ok I see. this seems a bit out of the realm of a "structure test" to me, but you could probably accomplish it with the structure test framework. don't think there's anything stopping you from just issuing the compile command through a command test (probably as a setup command), and then another command that actually calls the binary and checks the output to ensure it worked properly. I have plans to add support for checking the error code of a command test, which seems like an easy way for you guys to make sure your program worked correctly.

**Comment by odeke-em** [comment]
@broady some merge conflicts to be fixed.

**Comment by potjanat** [comment]
Dd

**Pull request closed without merging**